aws/resource_aws_sagemaker_notebook_instance.go

```diff
@@ -0,0 +1,390 @@
+		}
+
+		if _, err := conn.UpdateNotebookInstance(updateOpts); err != nil {
+			return fmt.Errorf("error updating sagemaker notebook instance: %s", err)
+		}
+
+		stateConf := &resource.StateChangeConf{
+			Pending: []string{
+				sagemaker.NotebookInstanceStatusUpdating,
+			},
+			Target:  []string{sagemaker.NotebookInstanceStatusStopped},
+			Refresh: sagemakerNotebookInstanceStateRefreshFunc(conn, d.Id()),
+			Timeout: 10 * time.Minute,
+		}
+		_, err := stateConf.WaitForState()
+		if err != nil {
+			return fmt.Errorf("error waiting for sagemaker notebook instance %q to update: %s", d.Id(), err)
+		}
+
+		// Restart if needed
+		if previousStatus == sagemaker.NotebookInstanceStatusInService {
+			startOpts := &sagemaker.StartNotebookInstanceInput{
+				NotebookInstanceName: aws.String(d.Id()),
+			}
+
+			// StartNotebookInstance sometimes doesn't take so we'll check for a state change and if
+			// it doesn't change we'll send another request
+			err := resource.Retry(5*time.Minute, func() *resource.RetryError {
+				if _, err := conn.StartNotebookInstance(startOpts); err != nil {
+					return resource.NonRetryableError(fmt.Errorf("error starting sagemaker notebook instance %q: %s", d.Id(), err))
+				}
+				stateConf := &resource.StateChangeConf{
+					Pending: []string{
+						sagemaker.NotebookInstanceStatusStopped,
+					},
+					Target:  []string{sagemaker.NotebookInstanceStatusInService, sagemaker.NotebookInstanceStatusPending},
+					Refresh: sagemakerNotebookInstanceStateRefreshFunc(conn, d.Id()),
+					Timeout: 30 * time.Second,
+				}
+				_, err := stateConf.WaitForState()
+				if err != nil {
+					return resource.RetryableError(fmt.Errorf("error waiting for sagemaker notebook instance %q to start: %s", d.Id(), err))
+				}
+
+				return nil
+			})
+			if err != nil {
+				return err
+			}
+
+			stateConf := &resource.StateChangeConf{
+				Pending: []string{
+					sagemaker.NotebookInstanceStatusUpdating,
+					sagemaker.NotebookInstanceStatusPending,
+					sagemaker.NotebookInstanceStatusStopped,
+				},
+				Target:  []string{sagemaker.NotebookInstanceStatusInService},
+				Refresh: sagemakerNotebookInstanceStateRefreshFunc(conn, d.Id()),
+				Timeout: 10 * time.Minute,
+			}
+			_, err = stateConf.WaitForState()
+			if err != nil {
+				return fmt.Errorf("error waiting for sagemaker notebook instance %q to start after update: %s", d.Id(), err)
+			}
+
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	d.Partial(false)
+
+	return resourceAwsSagemakerNotebookInstanceRead(d, meta)
+}
+
+func resourceAwsSagemakerNotebookInstanceDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).sagemakerconn
+
+	describeNotebookInput := &sagemaker.DescribeNotebookInstanceInput{
+		NotebookInstanceName: aws.String(d.Id()),
+	}
+	notebook, err := conn.DescribeNotebookInstance(describeNotebookInput)
+	if err != nil {
+		if sagemakerErr, ok := err.(awserr.Error); ok && sagemakerErr.Code() == "ResourceNotFound" {
```

Comment:
Ah ha -- this is related to above ResourceNotFound comment. All instances of this logic should likely be:

```suggestion
		if isAWSErr(err, sagemaker.ErrCodeResourceNotFound, "") {
```

Reply:
This is a mistake and it needs to be like the above as RecordNotFound